server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -4650,16 +4650,25 @@ private void updateVMDiskController(UserVmVO vm, Map<String, String> customParam
                 vm.setDetail(VmDetailConstants.DATA_DISK_CONTROLLER, dataDiskControllerSetting);
             }
 
-            String controllerSetting = StringUtils.defaultIfEmpty(_configDao.getValue(Config.VmwareRootDiskControllerType.key()),
-                    Config.VmwareRootDiskControllerType.getDefaultValue());
-
             // Don't override if VM already has root/data disk controller detail
             if (vm.getDetail(VmDetailConstants.ROOT_DISK_CONTROLLER) == null) {
-                vm.setDetail(VmDetailConstants.ROOT_DISK_CONTROLLER, controllerSetting);
+                String vmwareRootDiskControllerTypeFromSetting = StringUtils.defaultIfEmpty(_configDao.getValue(Config.VmwareRootDiskControllerType.key()),
+                        Config.VmwareRootDiskControllerType.getDefaultValue());
+                vm.setDetail(VmDetailConstants.ROOT_DISK_CONTROLLER, vmwareRootDiskControllerTypeFromSetting);
+                if (vm.getDetail(VmDetailConstants.DATA_DISK_CONTROLLER) == null) {
+                    if (vmwareRootDiskControllerTypeFromSetting.equalsIgnoreCase("scsi")) {
+                        vm.setDetail(VmDetailConstants.DATA_DISK_CONTROLLER, "scsi");
+                    }
+                }
```

Comment:
Are those lines necessary?

In the condition below, if the `dataDiskController` detail is null, and the `rootDiskController` is a SCSI controller, we're already updating `dataDiskController` to match `rootDiskController`.

Reply:
ah, forgot to remove them. thanks for the review @winterhazel, will update.